Added pagination and first level caching to per_app_list

nikolasgianna opened the pull request:
Implemented pagination and first level caching in per_app_list.php

AenBleidd commented:
@davidpanderson Please take a look at this PR. It seems that it is similar to the changes you've made in  ee3f8f1aa232a615f40c2dbee9047e2fb6551b83

davidpanderson commented:
Oops!  I didn't see this before my commit, which does essentially the same thing.

lfield commented:
Please note that this also implements caching which is also required.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -96,34 +107,72 @@ function show_row($item, $apps, $is_team, $i) {
+if (!$offset) $offset=0;
+if ($offset % $items_per_page) $offset = 0;
+$x = $is_team?"teams":"participants";
+page_head("Top $x by application");
```
Can you make this translatable? Something like `tra("Top %1 by application", $x)` should work.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -96,34 +107,72 @@ function show_row($item, $apps, $is_team, $i) {
+    echo "<a href=per_app_list.php?appid=$appid&is_team=$is_team&is_total=$is_total&offset=$new_offset>".tra("Previous %1", $items_per_page)."</a> &middot; ";
+}
+
+if ($n==$items_per_page) {
```
Please put spaces around the `==` operator.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -63,13 +67,20 @@ function show_header($is_team, $apps, $appid, $is_total) {
 // show a user or team, with their credit for each app
 //
 function show_row($item, $apps, $is_team, $i) {
```
There is still a DB lookup in this function which is not cached.

nikolasgianna replied:
The reason first level caching was mentioned on the title of the pull request is because only the choice of items and the chosen app credit for each of them is cached. A DB lookup still has to be done for the credit of the other apps' credit and caching that too would be a possible future improvement.

ChristianBeer replied:
I understand. But then the list will not be sorted. The sorting is done in the query that gets cached but the actual value is taken directly from the DB. This will lead to users that have a higher value listed after users with a lower value. This is perceived as a bug by the user. So I think we should cache everything right away in `get_top_items()`.

nikolasgianna replied:
Ok, so I will look into implementing full caching. Are there any suggestions or ideas on how to best do so?

ChristianBeer replied:
Rough sketch: Loop over the array returned by `get_top_items` and query credit and RAC per app and add this to `$data` (so it ends up in `$item` later). Then cache `$data`. Adjust the `foreach ($apps as $app) {` loop in `show_row` to not query the DB but look in `$item` for credit and RAC.

davidpanderson commented:
We didn't have the right indices on the credit_team and credit_user tables, so queries for high-score items were slow.  I added the indices.  With this change, caching may not be necessary.

nikogianna commented:
I added full caching and some checks in case there is some entry in the credit_user or credit_team table that doesn't correspond to an entry in the user or team table respectively.

nikolasgianna commented:
Implemented full caching and added checks in case an entry in the credit_team or credit_user tables doesn't correspond to an entry in the team or user tables respectively.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -56,74 +60,319 @@ function show_header($is_team, $apps, $appid, $is_total) {
+
+}
+
+function choose_users($x, $appid, $used_offset, $up) {
```
This and `choose_teams()` is compensating for the lack of a proper SQL JOIN clause implemented in `BoincCreditUser` for this use-case with several DB calls to eliminate old entries (mainly for the teams). Thus it would make more sense to me to write a custom `BoincCreditUser::get_list($where_clause, $order_clause, $limit)` function which does something like 
```
SELECT * from user u join user_credit cu on u.id=cu.userid where $where_clause order by $order_clause limit $limit
```
This will eliminate the need to check if a user or team still exists (because that is not part of the returned list) reducing the amount of queries per call.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -56,74 +60,319 @@ function show_header($is_team, $apps, $appid, $is_total) {
+     foreach ($data as $item) {
+                $team = BoincTeam::lookup_id($item->teamid);
+                if (is_object($team)) {
+                   $sign = "<td>".team_links($team)."</td>";
```
Please don't store HTML in the cache. I know BOINC is not separating Logic and Presentation very good but HTML belongs into the `show_row()` function. The cache should only contain the raw values. This has the advantage that you don't need to invalidate the whole cache just because you tweaked the output a little bit.

In this occasion it seems ok to cache the output from `team_links()` (without the surrounding td) which is HTML just to avoid the Badge DB lookup when displaying a cached page. But for the other values I don't see the need to store anything but the raw values.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -56,74 +60,319 @@ function show_header($is_team, $apps, $appid, $is_total) {
+      //
+     if (!$up) {
+      while ($check != $fetch) {
+             $data = BoincCreditUser::enum("appid=$appid order by $x desc limit $used_offset,$fetched_items");
```
This looks like tab indentation. BOINCs Coding style for this part of the source code is to use 4 spaces per indentation level. Please adjust your editor accordingly.

nikolasgianna commented:
I considered the DB layer to be out of bounds trying to touch as little as possible and aiming at doing everything with the given functions. Since the suggestion is that I implement a new function I have tried to follow the pattern and have overridden a base class function that can be reused.

davidpanderson commented:
With the addition of the indices on credit_user and credit_team,
performance may be OK without caching or joins.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -26,104 +26,243 @@
-    if ($app->id == $appid && ($is_total?$i:!$i)) {
+function col_title($is_team, $app, $appid, $is_total, $i)
+{
+    $x = $i ? "Total" : "Average";
```
If this is shown to the user you need to enclose it with `tra()`

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -26,104 +26,243 @@
         return $x;
     } else {
-        return "<a href=per_app_list.php?appid=$app->id&is_team=$is_team&is_total=$i>$x</a>";
+        return "<a href=per_app_list_dev.php?appid=$app->id&is_team=$is_team&is_total=$i>$x</a>";
```
leftover from testing?

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -26,104 +26,243 @@
     if ($is_team) {
-        $items = BoincCreditTeam::enum("appid=$appid order by $x desc");
+        
+        $data  = BoincCreditTeam::show_list("appid=$appid", $x, $offset . ", " . $items_per_page);
```
I personally associate functions that are called `show_` with ones that `echo` stuff and functions that return data are usually called `get_`. So I find this a bit misleading.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -26,104 +26,243 @@
+    $offset = 0;
+
+$x = $is_team ? "teams" : "participants";
+page_head(tra("Top %1 by application", $x));
```
To give the translator a chance to translate this better I guess there should be two complete strings without the `%1`. like this:
```
$title= $is_team ? tra("Top teams by application") : tra("Top participants by application");
page_head($title);
```

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -26,104 +26,243 @@
+
+if ($offset > 0) {
+    $new_offset = $offset - $items_per_page;
+    echo "<a href=per_app_list_dev.php?appid=$appid&is_team=$is_team&is_total=$is_total&offset=$new_offset>" . tra("Previous %1", $items_per_page) . "</a> &middot; ";
```
another left over from development. Also 5 lines down too.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -26,104 +26,243 @@
+    "appid",
+    "is_total",
+    "added_offset",
+    "up",
```
We don't need `added_offset` and `up` anymore.

ChristianBeer commented on html/user/per_app_list.php:
```diff
@@ -26,104 +26,243 @@
-
+    
     echo "<tr>";
     echo "<th>Rank</th><th>Name</th>\n";
```
Some missing `tra()`'s not spotted earlier.

nikolasgianna commented:
Using microtime() I am getting 650 microsecs on average for building the table from JOIN queries and 750 microsecs on average without the JOIN . Interestingly, using first level caching only, the table takes 80 millisecs on average to build and without any caching at all it takes 100 millisecs. These are all quick measurements just to get a rough idea of the times.

ChristianBeer commented:
The timings are not that promising but we recognized that it is better to have one slower join query than lots of faster select queries. At one point the amount of fast queries becomes the bigger problem which bogs down the whole database.

nikolasgianna commented:
Hope I have not missed anything.

ChristianBeer commented:
In f041c5873aee0c0e59b518962a8db69c1f61b658 you introduced tabs again where there were spaces before.

nikolasgianna commented:
Sorry for that.

davidpanderson commented:
alter table credit_user
    add index cu_total(appid, total),
    add index cu_avg(appid, expavg);

alter table credit_team
    add index ct_total(appid, total),
    add index ct_avg(appid, expavg);

Please add these indices and check the performance.

nikolasgianna commented:
The indices have been added. The performance looks good.

ChristianBeer commented:
@davidpanderson I would suggest to revert ee3f8f1aa232a615f40c2dbee9047e2fb6551b83 completely and re-implement the other changes (inc/bootstrap.inc) in a separate commit.

@nikolasgianna Can you add the CSS class changes from lines 49 and 58 of ee3f8f1aa232a615f40c2dbee9047e2fb6551b83 to your commit? That would completely substitute the relevant portions of the commit.

ChristianBeer commented:
Thanks for your patience.

nikolasgianna commented:
Thank you for the feedback.
